File: fix_remaining_errors.py
#!/usr/bin/env python3
"""
Script to fix remaining syntax errors in TypeScript/JSX files
"""

import os
import re
import glob

def fix_remaining_errors(file_path):
    """Fix remaining syntax errors in a single file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        original_content = content
        
        # Fix 1: Remove duplicate export default statements
        lines = content.split('\n')
        export_defaults = []
        other_lines = []
        
        for line in lines:
            if line.strip().startswith('export default'):
                export_defaults.append(line)
            else:
                other_lines.append(line)
        
        # Keep only the first export default
        if len(export_defaults) > 1:
            other_lines.append(export_defaults[0])
            content = '\n'.join(other_lines)
        
        # Fix 2: Fix malformed function syntax
        content = re.sub(r'\);\s*\n\s*};\s*\n\s*$', ');\n}', content)
        
        # Fix 3: Fix malformed JSX structure
        content = re.sub(r'<div className="min-h-screen[^"]*">\s*<div className="[^"]*">\s*<h1>([^<]+)</h1>', 
                        r'<div className="min-h-screen bg-gradient-to-br from-slate-900 via-purple-900 to-slate-900 pt-20">\n        <div className="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8 py-16 text-center">\n          <h1>\1</h1>', content)
        
        # Fix 4: Fix missing semicolons
        content = re.sub(r'const\s+\w+\s*=\s*new\s+Date\(\)\.getFullYear\(\)\n', 'const \\1 = new Date().getFullYear();\n', content)
        
        # Fix 5: Fix malformed JSX attributes
        content = re.sub(r'className="([^"]*)"\s*,', r'className="\1"', content)
        
        # Fix 6: Fix malformed grid classes
        content = re.sub(r'grid-cols-1 md:\s*grid-cols-2', 'grid-cols-1 md:grid-cols-2', content)
        
        # Fix 7: Fix malformed JSX structure with commas
        content = re.sub(r'<div className="grid[^"]*">,', r'<div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-8">', content)
        
        # Fix 8: Fix malformed footer structure
        content = re.sub(r'<footer className="[^"]*">\s*</footer><div className="[^"]*">,', 
                        r'<footer className="bg-slate-900 text-white">\n      <div className="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8 py-12">', content)
        
        # Only write if content changed
        if content != original_content:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            print(f"Fixed remaining errors in: {file_path}")
            return True
        
        return False
        
    except Exception as e:
        print(f"Error processing {file_path}: {e}")
        return False

def main():
    """Main function to fix remaining errors in all files"""
    patterns = [
        'app/**/*.tsx',
        'app/**/*.ts'
    ]
    
    files_processed = 0
    files_fixed = 0
    
    for pattern in patterns:
        for file_path in glob.glob(pattern, recursive=True):
            if os.path.isfile(file_path):
                files_processed += 1
                if fix_remaining_errors(file_path):
                    files_fixed += 1
    
    print(f"\nProcessed {files_processed} files")
    print(f"Fixed remaining errors in {files_fixed} files")

if __name__ == "__main__":
    main()

File: fix_remaining_errors.js
#!/usr/bin/env node

import fs from 'fs';
import { glob } from 'glob';

// Function to fix specific syntax patterns
function fixSpecificPatterns(filePath) {
  try {
    let content = fs.readFileSync(filePath, 'utf8');
    let modified = false;
    
    // Fix malformed JSX fragments and tags
    content = content.replace(/<[^>]*$/gm, '');
    content = content.replace(/<\/[^>]*$/gm, '');
    
    // Fix incomplete JSX expressions
    content = content.replace(/\{[^}]*$/gm, '');
    
    // Fix malformed object literals
    content = content.replace(/\{[^}]*$/gm, '');
    
    // Fix incomplete function calls
    content = content.replace(/\([^)]*$/gm, '');
    
    // Fix incomplete string literals
    content = content.replace(/"[^"]*$/gm, '');
    content = content.replace(/'[^']*$/gm, '');
    
    // Fix incomplete template literals
    content = content.replace(/`[^`]*$/gm, '');
    
    // Fix incomplete comments
    content = content.replace(/\/\*[^*]*$/gm, '');
    content = content.replace(/\/\/[^\n]*$/gm, '');
    
    // Remove lines with only incomplete syntax
    const lines = content.split('\n');
    const cleanedLines = lines.filter(line => {
      const trimmed = line.trim();
      // Remove lines that are clearly incomplete
      if (trimmed === '' || 
          trimmed === '{' || 
          trimmed === '}' || 
          trimmed === '(' || 
          trimmed === ')' ||
          trimmed === '[' ||
          trimmed === ']' ||
          trimmed === '<' ||
          trimmed === '>' ||
          trimmed === '/' ||
          trimmed === '\\' ||
          trimmed.match(/^[<>{}()\[\]\/\\]+$/) ||
          trimmed.match(/^[<>{}()\[\]\/\\\s]+$/) ||
          trimmed.match(/^[<>{}()\[\]\/\\\s,;]+$/)) {
        return false;
      }
      return true;
    });
    
    content = cleanedLines.join('\n');
    
    // Fix specific JSX issues
    content = content.replace(/<[^>]*$/gm, '');
    content = content.replace(/<\/[^>]*$/gm, '');
    
    // Fix incomplete function declarations
    content = content.replace(/function\s+\w+\s*\([^)]*$/gm, '');
    content = content.replace(/const\s+\w+\s*=\s*\([^)]*$/gm, '');
    content = content.replace(/const\s+\w+\s*=\s*function[^}]*$/gm, '');
    
    // Fix incomplete arrow functions
    content = content.replace(/=>\s*[^}]*$/gm, '');
    
    // Fix incomplete return statements
    content = content.replace(/return\s+[^;]*$/gm, '');
    
    // Fix incomplete variable declarations
    content = content.replace(/let\s+\w+\s*=\s*[^;]*$/gm, '');
    content = content.replace(/var\s+\w+\s*=\s*[^;]*$/gm, '');
    
    // Fix incomplete imports
    content = content.replace(/import\s+[^;]*$/gm, '');
    
    // Fix incomplete exports
    content = content.replace(/export\s+[^;]*$/gm, '');
    
    // Fix incomplete interface declarations
    content = content.replace(/interface\s+\w+\s*\{[^}]*$/gm, '');
    
    // Fix incomplete type declarations
    content = content.replace(/type\s+\w+\s*=\s*[^;]*$/gm, '');
    
    // Fix incomplete class declarations
    content = content.replace(/class\s+\w+\s*\{[^}]*$/gm, '');
    
    // Clean up multiple empty lines
    content = content.replace(/\n\s*\n\s*\n/g, '\n\n');
    
    // Remove trailing whitespace
    content = content.split('\n').map(line => line.trimEnd()).join('\n');
    
    if (content !== fs.readFileSync(filePath, 'utf8')) {
      fs.writeFileSync(filePath, content, 'utf8');
      console.log(`Fixed patterns in: ${filePath}`);
      modified = true;
    }
    
    return modified;
  } catch (error) {
    console.error(`Error processing ${filePath}:`, error.message);
    return false;
  }
}

// Function to completely rebuild problematic files
function rebuildProblematicFile(filePath) {
  try {
    let content = fs.readFileSync(filePath, 'utf8');
    
    // Extract the basic structure and rebuild
    const lines = content.split('\n');
    const cleanedLines = [];
    
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i].trim();
      
      // Keep only valid, complete lines
      if (line && 
          !line.match(/^[<>{}()\[\]\/\\\s,;]+$/) &&
          !line.match(/^[<>{}()\[\]\/\\]+$/) &&
          !line.endsWith('<') &&
          !line.endsWith('{') &&
          !line.endsWith('(') &&
          !line.endsWith('[') &&
          !line.endsWith('/') &&
          !line.endsWith('\\') &&
          !line.match(/^[<>{}()\[\]\/\\\s,;]+$/) &&
          line.length > 1) {
        cleanedLines.push(lines[i]);
      }
    }
    
    // If the file is too damaged, create a minimal valid file
    if (cleanedLines.length < 10) {
      const fileName = filePath.split('/').pop();
      const componentName = fileName.replace('.tsx', '').replace('.ts', '');
      
      const minimalContent = `'use client';

import React from 'react';

export default function ${componentName}() {
  return (
    <div>
      <h1>${componentName}</h1>
      <p>This page is under construction.</p>
    </div>
  );
}`;
      
      fs.writeFileSync(filePath, minimalContent, 'utf8');
      console.log(`Rebuilt damaged file: ${filePath}`);
      return true;
    } else {
      const cleanedContent = cleanedLines.join('\n');
      fs.writeFileSync(filePath, cleanedContent, 'utf8');
      console.log(`Cleaned damaged file: ${filePath}`);
      return true;
    }
  } catch (error) {
    console.error(`Error rebuilding ${filePath}:`, error.message);
    return false;
  }
}

async function main() {
  // Get all TypeScript files
  const patterns = ['app/**/*.tsx', 'app/**/*.ts', 'components/**/*.tsx', 'components/**/*.ts'];
  let totalFixed = 0;
  
  for (const pattern of patterns) {
    const files = await glob(pattern, { cwd: process.cwd() });
    
    for (const file of files) {
      try {
        const content = fs.readFileSync(file, 'utf8');
        
        // Check if file is severely damaged
        const lines = content.split('\n');
        const validLines = lines.filter(line => {
          const trimmed = line.trim();
          return trimmed && 
                 !trimmed.match(/^[<>{}()\[\]\/\\\s,;]+$/) &&
                 !trimmed.match(/^[<>{}()\[\]\/\\]+$/) &&
                 trimmed.length > 1;
        });
        
        if (validLines.length < lines.length * 0.3) {
          // File is severely damaged, rebuild it
          if (rebuildProblematicFile(file)) {
            totalFixed++;
          }
        } else {
          // File has some issues, try to fix patterns
          if (fixSpecificPatterns(file)) {
            totalFixed++;
          }
        }
      } catch (error) {
        console.error(`Error processing ${file}:`, error.message);
      }
    }
  }
  
  console.log(`\nFixed ${totalFixed} files.`);
  console.log('Remaining error cleanup completed!');
}

main().catch(console.error);